File: app/src/main/java/com/dream/countryinfo/network/CountryApiHelper.java
package com.dream.countryinfo.network;

import android.support.annotation.NonNull;

import com.dream.countryinfo.CountryApp;
import com.dream.countryinfo.feature.country.CountryDetail;

import java.lang.ref.SoftReference;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Created by lixingming on 12/11/2017.
 */

public class CountryApiHelper {

    private List<SoftReference<Call<List<Map<String, String>>>>> listRefCall;

    private CountryApi countryApi;

    private volatile static CountryApiHelper singleton;

    private CountryApiHelper (){}

    public static CountryApiHelper getSingleton() {
        if (singleton == null) {
            synchronized (CountryApiHelper.class) {
                if (singleton == null) {
                    singleton = new CountryApiHelper();
                    CountryApiCreator apiCreator = new CountryApiCreator(CountryApp.getApplication()
                            .getApplicationContext());
                    singleton.countryApi = apiCreator.createApi("https://restcountries.eu/");

                    singleton.listRefCall = new ArrayList<>();
                }
            }
        }
        return singleton;
    }

    public void searchCountriesByName(String searchText, String fields, MyCallback<List<Map<String, String>>>
            callback) {
        Call<List<Map<String, String>>> call = countryApi.searchCountriesByName(searchText, fields);
        call.enqueue(new DefaultCallback<>(callback));

        listRefCall.add(new SoftReference<>(call));
    }

    public void getCountriesByName(String countryName, String fields, MyCallback<List<CountryDetail>> callback) {
        Call<List<CountryDetail>> call = countryApi.getCountriesByName(countryName, fields);
        call.enqueue(new DefaultCallback<>(callback));
    }

    public void cancelSearchCountryCalls() {
        for (SoftReference<Call<List<Map<String, String>>>> ref : listRefCall) {
            if (ref.get() != null) {
                ref.get().cancel();
            }
        }
        listRefCall.clear();
    }

    class DefaultCallback<T> implements Callback<T> {
        private MyCallback<T> callback;

        public DefaultCallback(MyCallback<T> callback) {
            this.callback = callback;
        }

        @Override
        public void onResponse(@NonNull Call<T> call, @NonNull Response<T> response) {
            if (!call.isCanceled())
                callback.onResponse(response.body());
        }

        @Override
        public void onFailure(@NonNull Call<T> call, @NonNull Throwable t) {
            if (!call.isCanceled())
                callback.onFailure(t);
        }
    }

    public interface MyCallback<T> {
        // this callback just ignore the call
        void onResponse(T responseData);

        // this callback just ignore the call
        void onFailure(Throwable t);
    }
}

File: app/src/main/java/com/dream/countryinfo/network/CountryApiHelper.kt
package com.dream.countryinfo.network

import com.dream.countryinfo.CountryApp
import com.dream.countryinfo.feature.country.CountryDetail

import java.lang.ref.SoftReference
import java.util.ArrayList

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/**
 * Created by lixingming on 12/11/2017.
 */

class CountryApiHelper private constructor() {

    private var listRefCall: MutableList<SoftReference<Call<List<Map<String, String>>>>> =
            ArrayList<SoftReference<Call<List<Map<String, String>>>>>().toMutableList()

    private var countryApi: CountryApi? = null

    fun searchCountriesByName(searchText: String, fields: String, callback: MyCallback<List<Map<String, String>>>) {
        val call = countryApi!!.searchCountriesByName(searchText, fields)
        call.enqueue(DefaultCallback(callback))

        listRefCall.add(SoftReference(call))
    }

    fun getCountriesByName(countryName: String, fields: String, callback: MyCallback<List<CountryDetail>>) {
        val call = countryApi!!.getCountriesByName(countryName, fields)
        call.enqueue(DefaultCallback(callback))
    }

    fun cancelSearchCountryCalls() {
        for (ref in listRefCall) {
            if (ref.get() != null) {
                (ref.get() as Call<List<Map<String, String>>>).cancel()
            }
        }
        listRefCall.clear()
    }

    internal inner class DefaultCallback<T>(private val callback: MyCallback<T>) : Callback<T> {

        override fun onResponse(call: Call<T>, response: Response<T>) {
            if (!call.isCanceled)
                callback.onResponse(response.body())
        }

        override fun onFailure(call: Call<T>, t: Throwable) {
            if (!call.isCanceled)
                callback.onFailure(t)
        }
    }

    interface MyCallback<T> {
        // this callback just ignore the call
        fun onResponse(responseData: T?)

        // this callback just ignore the call
        fun onFailure(t: Throwable)
    }

    companion object {

        @Volatile private var instance: CountryApiHelper? = null

        fun getSingleton(): CountryApiHelper {
            if (instance == null) {
                synchronized(CountryApiHelper::class.java) {
                    if (instance == null) {
                        instance = CountryApiHelper()
                        val apiCreator = CountryApiCreator(CountryApp.application!!
                                .getApplicationContext())
                        instance!!.countryApi = apiCreator.createApi("https://restcountries.eu/")

//                        instance!!.listRefCall = ArrayList()
                    }
                }
            }
            return instance!!
        }
    }
}
